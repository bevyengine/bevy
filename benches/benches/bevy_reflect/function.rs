use bevy_reflect::func::{ArgList, IntoFunction, TypedFunction};
use bevy_reflect::prelude::*;
use criterion::{criterion_group, criterion_main, BatchSize, Criterion};

criterion_group!(benches, typed, into, call, clone);
criterion_main!(benches);

fn add(a: i32, b: i32) -> i32 {
    a + b
}

fn typed(c: &mut Criterion) {
    c.benchmark_group("typed")
        .bench_function("function", |b| {
            b.iter(|| add.get_function_info());
        })
        .bench_function("closure", |b| {
            let capture = 25;
            let closure = |a: i32| a + capture;
            b.iter(|| closure.get_function_info());
        });
}

fn into(c: &mut Criterion) {
    c.benchmark_group("into")
        .bench_function("function", |b| {
            b.iter(|| add.into_function());
        })
        .bench_function("closure", |b| {
            let capture = 25;
            let closure = |a: i32| a + capture;
            b.iter(|| closure.into_function());
        });
}

fn call(c: &mut Criterion) {
    c.benchmark_group("call")
        .bench_function("function", |b| {
            let add = add.into_function();
            b.iter_batched(
                || ArgList::new().push_owned(75_i32).push_owned(25_i32),
                |args| add.call(args),
                BatchSize::SmallInput,
            );
        })
        .bench_function("closure", |b| {
            let capture = 25;
            let add = (|a: i32| a + capture).into_function();
            b.iter_batched(
                || ArgList::new().push_owned(75_i32),
                |args| add.call(args),
                BatchSize::SmallInput,
            );
        });
}

fn clone(c: &mut Criterion) {
    c.benchmark_group("clone").bench_function("function", |b| {
        let add = add.into_function();
        b.iter(|| add.clone());
    });
}
