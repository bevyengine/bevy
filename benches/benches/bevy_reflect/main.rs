use criterion::criterion_main;

mod function;
mod list;
mod map;
mod path;
mod r#struct;

criterion_main!(
    function::benches,
    list::benches,
    map::benches,
    path::benches,
    r#struct::benches,
);
