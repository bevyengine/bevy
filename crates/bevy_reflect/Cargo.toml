[package]
name = "bevy_reflect"
version = "0.4.0"
edition = "2018"
authors = [
    "Bevy Contributors <bevyengine@gmail.com>",
    "Carter Anderson <mcanders1@gmail.com>",
]
description = "Dynamically interact with rust types"
homepage = "https://bevyengine.org"
repository = "https://github.com/bevyengine/bevy"
license = "MIT"
keywords = ["bevy"]
readme = "README.md"

[features]
bevy = ["bevy_ecs", "bevy_app", "glam", "smallvec"]

[dependencies]
# bevy
bevy_reflect_derive = { path = "bevy_reflect_derive", version = "0.4.0" }
bevy_utils = { path = "../bevy_utils", version = "0.4.0" }
bevy_ecs = { path = "../bevy_ecs", version = "0.4.0", optional = true }
bevy_app = { path = "../bevy_app", version = "0.4.0", optional = true }

# other
erased-serde = "0.3"
downcast-rs = "1.2"
parking_lot = "0.11.0"
thiserror = "1.0"
serde = "1"
smallvec = { version = "1.4", features = ["serde"], optional = true }
glam = { version = "0.13.0", features = ["serde"], optional = true }

[dev-dependencies]
ron = "0.6.2"
