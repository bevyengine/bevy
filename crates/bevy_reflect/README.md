# Bevy Reflect

This crate enables you to dynamically interact with Rust types:

* Derive the Reflect traits
* Interact with fields using their names (for named structs) or indices (for tuple structs)
* "Patch" your types with new values
* Look up nested fields using "path strings"
* Iterate over struct fields
* Automatically serialize and deserialize via Serde (without explicit serde impls)
* Trait "reflection" 

## Features

### Derive the Reflect traits

```rust
// this will automatically implement the Reflect trait and the Struct trait (because the type is a struct)
#[derive(Reflect)]
struct Foo {
    a: u32,
    b: Bar
    c: Vec<i32>,
    d: Vec<Bar>,
}

// this will automatically implement the Reflect trait and the TupleStruct trait (because the type is a tuple struct)
#[derive(Reflect)]
struct Bar(String);

#[derive(Reflect)]
struct Baz {
    value: f32,
};

// We will use this value to illustrate `bevy_reflect` features
let mut foo = Foo {
    a: 1,
    b: Bar("hello".to_string()),
    c: vec![1, 2]
    d: vec![Baz { value: 3.14 }]
};
```

### Interact with fields using their names

```rust
assert_eq!(*foo.get_field::<u32>("a").unwrap(), 1);

*foo.get_field_mut::<u32>("a").unwrap() = 2;

assert_eq!(foo.a, 2);
```

### "Patch" your types with new values

```rust
let mut dynamic_struct = DynamicStruct::default();
dynamic_struct.insert("a", 42u32);
dynamic_struct.insert("c", vec![3, 4, 5]);

foo.apply(&dynamic_struct);

assert_eq!(foo.a, 42);
assert_eq!(foo.c, vec![3, 4, 5]);
```

### Look up nested fields using "path strings"

```rust
let value = *foo.get_path::<f32>("d[0].value").unwrap();
assert_eq!(value, 3.14);
```

### Iterate over struct fields

```rust
for (i, value: &Reflect) in foo.iter_fields().enumerate() {
    let field_name = foo.name_at(i).unwrap();
    if let Ok(value) = value.downcast_ref::<u32>() {
        println!("{} is a u32 with the value: {}", field_name, *value);
    } 
}
```

### Automatically serialize and deserialize via Serde (without explicit serde impls)

```rust
let mut registry = TypeRegistry::default();
registry.register::<u32>();
registry.register::<i32>();
registry.register::<f32>();
registry.register::<String>();
registry.register::<Bar>();
registry.register::<Baz>();

let serializer = ReflectSerializer::new(&foo, &registry);
let serialized = ron::ser::to_string_pretty(&serializer, ron::ser::PrettyConfig::default()).unwrap();

let mut deserializer = ron::de::Deserializer::from_str(&serialized).unwrap();
let reflect_deserializer = ReflectDeserializer::new(&registry);
let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
let dynamic_struct = value.take::<DynamicStruct>().unwrap();

assert!(foo.partial_eq(&dynamic_struct).unwrap());
```

### Trait "reflection"

Call a trait on a given &dyn Reflect reference without knowing the underlying type!

```rust
#[derive(Reflect)]
#[reflect(DoThing)]
struct MyType {
    value: String,
}

impl DoThing for MyType {
    fn do_thing(&self) -> String {
        format!("{} World!", self.value)
    }
}

#[reflect_trait]
pub trait DoThing {
    fn do_thing(&self) -> String;
}

// First, lets box our type as a Box<dyn Reflect>
let reflect_value: Box<dyn Reflect> = Box::new(MyType {
    value: "Hello".to_string(),
});

// This means we no longer have direct access to MyType or its methods. We can only call Reflect methods on reflect_value.
// What if we want to call `do_thing` on our type? We could downcast using reflect_value.downcast_ref::<MyType>(), but what if we
// don't know the type at compile time?

// Normally in rust we would be out of luck at this point. Lets use our new reflection powers to do something cool!
let mut type_registry = TypeRegistry::default()
type_registry.register::<MyType>();

// The #[reflect] attribute we put on our DoThing trait generated a new `ReflectDoThing` struct, which implements TypeData.
// This was added to MyType's TypeRegistration.
let reflect_do_thing = type_registry
    .get_type_data::<ReflectDoThing>(reflect_value.type_id())
    .unwrap();

// We can use this generated type to convert our `&dyn Reflect` reference to a `&dyn DoThing` reference
let my_trait: &dyn DoThing = reflect_do_thing.get(&*reflect_value).unwrap();

// Which means we can now call do_thing(). Magic!
println!("{}", my_trait.do_thing());

// This works because the #[reflect(MyTrait)] we put on MyType informed the Reflect derive to insert a new instance
// of ReflectDoThing into MyType's registration. The instance knows how to cast &dyn Reflect to &dyn MyType, because it
// knows that &dyn Reflect should first be downcasted to &MyType, which can then be safely casted to &dyn MyType
```

## Why make this?

The whole point of Rust is static safety! Why build something that makes it easy to throw it all away?

* Some problems are inherently dynamic (scripting, some types of serialization / deserialization)
* Sometimes the dynamic way is easier
* Sometimes the dynamic way puts less burden on your users to derive a bunch of traits (this was a big motivator for the Bevy project)