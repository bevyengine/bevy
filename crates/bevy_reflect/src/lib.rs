mod list;
mod map;
mod path;
mod reflect;
mod struct_trait;
mod tuple;
mod tuple_struct;
mod type_registry;
mod type_uuid;
mod impls {
    #[cfg(feature = "glam")]
    mod glam;
    #[cfg(feature = "smallvec")]
    mod smallvec;
    mod std;

    #[cfg(feature = "glam")]
    pub use self::glam::*;
    #[cfg(feature = "smallvec")]
    pub use self::smallvec::*;
    pub use self::std::*;
}

pub mod serde;
pub mod prelude {
    pub use crate::{
        reflect_trait, GetField, GetTupleStructField, Reflect, ReflectDeserialize, Struct,
        TupleStruct,
    };
}

pub use impls::*;
pub use list::*;
pub use map::*;
pub use path::*;
pub use reflect::*;
pub use struct_trait::*;
pub use tuple::*;
pub use tuple_struct::*;
pub use type_registry::*;
pub use type_uuid::*;

pub use bevy_reflect_derive::*;
pub use erased_serde;

#[cfg(test)]
#[allow(clippy::blacklisted_name, clippy::approx_constant)]
mod tests {
    use ::serde::de::DeserializeSeed;
    use bevy_utils::HashMap;
    use ron::{
        ser::{to_string_pretty, PrettyConfig},
        Deserializer,
    };

    use crate::serde::{ReflectDeserializer, ReflectSerializer};

    use super::*;

    #[test]
    fn reflect_struct() {
        #[derive(Reflect)]
        struct Foo {
            a: u32,
            b: f32,
            c: Bar,
        }
        #[derive(Reflect)]
        struct Bar {
            x: u32,
        }

        let mut foo = Foo {
            a: 42,
            b: 3.14,
            c: Bar { x: 1 },
        };

        let a = *foo.get_field::<u32>("a").unwrap();
        assert_eq!(a, 42);

        *foo.get_field_mut::<u32>("a").unwrap() += 1;
        assert_eq!(foo.a, 43);

        let bar = foo.get_field::<Bar>("c").unwrap();
        assert_eq!(bar.x, 1);

        // nested retrieval
        let c = foo.field("c").unwrap();
        if let ReflectRef::Struct(value) = c.reflect_ref() {
            assert_eq!(*value.get_field::<u32>("x").unwrap(), 1);
        } else {
            panic!("Expected a struct.");
        }

        // patch Foo with a dynamic struct
        let mut dynamic_struct = DynamicStruct::default();
        dynamic_struct.insert("a", 123u32);
        dynamic_struct.insert("should_be_ignored", 456);

        foo.apply(&dynamic_struct);
        assert_eq!(foo.a, 123);
    }

    #[test]
    fn reflect_map() {
        #[derive(Reflect, Hash)]
        #[reflect(Hash)]
        struct Foo {
            a: u32,
            b: String,
        }

        let key_a = Foo {
            a: 1,
            b: "k1".to_string(),
        };

        let key_b = Foo {
            a: 1,
            b: "k1".to_string(),
        };

        let key_c = Foo {
            a: 3,
            b: "k3".to_string(),
        };

        let mut map = DynamicMap::default();
        map.insert(key_a, 10u32);
        assert_eq!(10, *map.get(&key_b).unwrap().downcast_ref::<u32>().unwrap());
        assert!(map.get(&key_c).is_none());
        *map.get_mut(&key_b).unwrap().downcast_mut::<u32>().unwrap() = 20;
        assert_eq!(20, *map.get(&key_b).unwrap().downcast_ref::<u32>().unwrap());
    }

    #[test]
    #[allow(clippy::blacklisted_name)]
    fn reflect_unit_struct() {
        #[derive(Reflect)]
        struct Foo(u32, u64);

        let mut foo = Foo(1, 2);
        assert_eq!(1, *foo.get_field::<u32>(0).unwrap());
        assert_eq!(2, *foo.get_field::<u64>(1).unwrap());

        let mut patch = DynamicTupleStruct::default();
        patch.insert(3u32);
        patch.insert(4u64);
        assert_eq!(3, *patch.field(0).unwrap().downcast_ref::<u32>().unwrap());
        assert_eq!(4, *patch.field(1).unwrap().downcast_ref::<u64>().unwrap());

        foo.apply(&patch);
        assert_eq!(3, foo.0);
        assert_eq!(4, foo.1);

        let mut iter = patch.iter_fields();
        assert_eq!(3, *iter.next().unwrap().downcast_ref::<u32>().unwrap());
        assert_eq!(4, *iter.next().unwrap().downcast_ref::<u64>().unwrap());
    }

    #[test]
    #[should_panic(expected = "the given key does not support hashing")]
    fn reflect_map_no_hash() {
        #[derive(Reflect)]
        struct Foo {
            a: u32,
        }

        let foo = Foo { a: 1 };

        let mut map = DynamicMap::default();
        map.insert(foo, 10u32);
    }

    #[test]
    fn reflect_ignore() {
        #[derive(Reflect)]
        struct Foo {
            a: u32,
            #[reflect(ignore)]
            _b: u32,
        }

        let foo = Foo { a: 1, _b: 2 };

        let values: Vec<u32> = foo
            .iter_fields()
            .map(|value| *value.downcast_ref::<u32>().unwrap())
            .collect();
        assert_eq!(values, vec![1]);
    }

    #[test]
    fn reflect_complex_patch() {
        #[derive(Reflect, Eq, PartialEq, Debug)]
        #[reflect(PartialEq)]
        struct Foo {
            a: u32,
            #[reflect(ignore)]
            _b: u32,
            c: Vec<isize>,
            d: HashMap<usize, i8>,
            e: Bar,
            f: (i32, Vec<isize>, Bar),
        }

        #[derive(Reflect, Eq, PartialEq, Debug)]
        #[reflect(PartialEq)]
        struct Bar {
            x: u32,
        }

        let mut hash_map = HashMap::default();
        hash_map.insert(1, 1);
        hash_map.insert(2, 2);
        let mut foo = Foo {
            a: 1,
            _b: 1,
            c: vec![1, 2],
            d: hash_map,
            e: Bar { x: 1 },
            f: (1, vec![1, 2], Bar { x: 1 }),
        };

        let mut foo_patch = DynamicStruct::default();
        foo_patch.insert("a", 2u32);
        foo_patch.insert("b", 2u32); // this should be ignored

        let mut list = DynamicList::default();
        list.push(3isize);
        list.push(4isize);
        list.push(5isize);
        foo_patch.insert("c", list.clone_dynamic());

        let mut map = DynamicMap::default();
        map.insert(2usize, 3i8);
        foo_patch.insert("d", map);

        let mut bar_patch = DynamicStruct::default();
        bar_patch.insert("x", 2u32);
        foo_patch.insert("e", bar_patch.clone_dynamic());

        let mut tuple = DynamicTuple::default();
        tuple.insert(2i32);
        tuple.insert(list);
        tuple.insert(bar_patch);
        foo_patch.insert("f", tuple);

        foo.apply(&foo_patch);

        let mut hash_map = HashMap::default();
        hash_map.insert(1, 1);
        hash_map.insert(2, 3);
        let expected_foo = Foo {
            a: 2,
            _b: 1,
            c: vec![3, 4, 5],
            d: hash_map,
            e: Bar { x: 2 },
            f: (2, vec![3, 4, 5], Bar { x: 2 }),
        };

        assert_eq!(foo, expected_foo);
    }

    #[test]
    fn reflect_serialize() {
        #[derive(Reflect)]
        struct Foo {
            a: u32,
            #[reflect(ignore)]
            _b: u32,
            c: Vec<isize>,
            d: HashMap<usize, i8>,
            e: Bar,
            f: String,
            g: (i32, Vec<isize>, Bar),
        }

        #[derive(Reflect)]
        struct Bar {
            x: u32,
        }

        let mut hash_map = HashMap::default();
        hash_map.insert(1, 1);
        hash_map.insert(2, 2);
        let foo = Foo {
            a: 1,
            _b: 1,
            c: vec![1, 2],
            d: hash_map,
            e: Bar { x: 1 },
            f: "hi".to_string(),
            g: (1, vec![1, 2], Bar { x: 1 }),
        };

        let mut registry = TypeRegistry::default();
        registry.register::<u32>();
        registry.register::<isize>();
        registry.register::<usize>();
        registry.register::<Bar>();
        registry.register::<String>();
        registry.register::<i8>();
        registry.register::<i32>();

        let serializer = ReflectSerializer::new(&foo, &registry);
        let serialized = to_string_pretty(&serializer, PrettyConfig::default()).unwrap();

        let mut deserializer = Deserializer::from_str(&serialized).unwrap();
        let reflect_deserializer = ReflectDeserializer::new(&registry);
        let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
        let dynamic_struct = value.take::<DynamicStruct>().unwrap();

        assert!(foo.reflect_partial_eq(&dynamic_struct).unwrap());
    }

    #[test]
    fn reflect_take() {
        #[derive(Reflect, Debug, PartialEq)]
        #[reflect(PartialEq)]
        struct Bar {
            x: u32,
        }

        let x: Box<dyn Reflect> = Box::new(Bar { x: 2 });
        let y = x.take::<Bar>().unwrap();
        assert_eq!(y, Bar { x: 2 });
    }

    #[test]
    fn dynamic_names() {
        let list = Vec::<usize>::new();
        let dyn_list = list.clone_dynamic();
        assert_eq!(dyn_list.type_name(), std::any::type_name::<Vec<usize>>());

        let map = HashMap::<usize, String>::default();
        let dyn_map = map.clone_dynamic();
        assert_eq!(
            dyn_map.type_name(),
            std::any::type_name::<HashMap<usize, String>>()
        );

        let tuple = (0usize, "1".to_string(), 2.0f32);
        let mut dyn_tuple = tuple.clone_dynamic();
        dyn_tuple.insert::<usize>(3);
        assert_eq!(
            dyn_tuple.type_name(),
            std::any::type_name::<(usize, String, f32, usize)>()
        );

        #[derive(Reflect)]
        struct TestStruct {
            a: usize,
        }
        let struct_ = TestStruct { a: 0 };
        let dyn_struct = struct_.clone_dynamic();
        assert_eq!(dyn_struct.type_name(), std::any::type_name::<TestStruct>());

        #[derive(Reflect)]
        struct TestTupleStruct(usize);
        let tuple_struct = TestTupleStruct(0);
        let dyn_tuple_struct = tuple_struct.clone_dynamic();
        assert_eq!(
            dyn_tuple_struct.type_name(),
            std::any::type_name::<TestTupleStruct>()
        );
    }
}
