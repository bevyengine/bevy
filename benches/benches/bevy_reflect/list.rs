use std::{iter, time::Duration};

use bevy_reflect::{DynamicList, List};
use criterion::{
    black_box, criterion_group, criterion_main, measurement::Measurement, BatchSize,
    BenchmarkGroup, BenchmarkId, Criterion, Throughput,
};

criterion_group!(
    benches,
    concrete_list_apply,
    concrete_list_clone_dynamic,
    dynamic_list_apply,
    dynamic_list_push
);
criterion_main!(benches);

const WARM_UP_TIME: Duration = Duration::from_millis(500);
const MEASUREMENT_TIME: Duration = Duration::from_secs(4);

// log10 scaling
const SIZES: [usize; 5] = [100_usize, 316, 1000, 3162, 10000];

fn list_apply<M, LBase, LPatch, F1, F2, F3>(
    group: &mut BenchmarkGroup<M>,
    bench_name: &str,
    f_base: F1,
    f_patch: F3,
) where
    M: Measurement,
    LBase: List,
    LPatch: List,
    F1: Fn(usize) -> F2,
    F2: Fn() -> LBase,
    F3: Fn(usize) -> LPatch,
{
    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));

        group.bench_with_input(
            BenchmarkId::new(bench_name, size),
            &size,
            |bencher, &size| {
                let f_base = f_base(size);
                let patch = f_patch(size);
                bencher.iter_batched(
                    || f_base(),
                    |mut base| base.apply(black_box(&patch)),
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

fn concrete_list_apply(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("concrete_list_apply");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    let empty_base = |_: usize| || Vec::<u64>::new();
    let full_base = |size: usize| move || iter::repeat(0).take(size).collect::<Vec<u64>>();
    let patch = |size: usize| iter::repeat(1).take(size).collect::<Vec<u64>>();

    list_apply(&mut group, "empty_base_concrete_patch", empty_base, patch);

    list_apply(&mut group, "empty_base_dynamic_patch", empty_base, |size| {
        patch(size).clone_dynamic()
    });

    list_apply(&mut group, "same_len_concrete_patch", full_base, patch);

    list_apply(&mut group, "same_len_dynamic_patch", full_base, |size| {
        patch(size).clone_dynamic()
    });

    group.finish();
}

fn concrete_list_clone_dynamic(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("concrete_list_clone_dynamic");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));

        group.bench_with_input(
            BenchmarkId::from_parameter(size),
            &size,
            |bencher, &size| {
                let v = iter::repeat(0).take(size).collect::<Vec<_>>();

                bencher.iter(|| black_box(&v).clone_dynamic());
            },
        );
    }

    group.finish();
}

fn dynamic_list_push(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("dynamic_list_push");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));

        group.bench_with_input(
            BenchmarkId::from_parameter(size),
            &size,
            |bencher, &size| {
                let src = iter::repeat(()).take(size).collect::<Vec<_>>();
                let dst = DynamicList::default();

                bencher.iter_batched(
                    || (src.clone(), dst.clone_dynamic()),
                    |(src, mut dst)| {
                        for item in src {
                            dst.push(item);
                        }
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }

    group.finish();
}

fn dynamic_list_apply(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("dynamic_list_apply");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    let empty_base = |_: usize| || Vec::<u64>::new().clone_dynamic();
    let full_base = |size: usize| move || iter::repeat(0).take(size).collect::<Vec<u64>>();
    let patch = |size: usize| iter::repeat(1).take(size).collect::<Vec<u64>>();

    list_apply(&mut group, "empty_base_concrete_patch", empty_base, patch);

    list_apply(&mut group, "empty_base_dynamic_patch", empty_base, |size| {
        patch(size).clone_dynamic()
    });

    list_apply(&mut group, "same_len_concrete_patch", full_base, patch);

    list_apply(&mut group, "same_len_dynamic_patch", full_base, |size| {
        patch(size).clone_dynamic()
    });

    group.finish();
}
