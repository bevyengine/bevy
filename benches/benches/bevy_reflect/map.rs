use std::{fmt::Write, iter, time::Duration};

use bevy_reflect::{DynamicMap, Map};
use bevy_utils::HashMap;
use criterion::{
    black_box, criterion_group, criterion_main, measurement::Measurement, BatchSize,
    BenchmarkGroup, BenchmarkId, Criterion, Throughput,
};

criterion_group!(
    benches,
    concrete_map_apply,
    dynamic_map_apply,
    dynamic_map_get,
    dynamic_map_insert
);
criterion_main!(benches);

const WARM_UP_TIME: Duration = Duration::from_millis(500);
const MEASUREMENT_TIME: Duration = Duration::from_secs(4);
const SIZES: [usize; 5] = [100, 316, 1000, 3162, 10000];

/// Generic benchmark for applying one `Map` to another.
///
/// `f_base` is a function which takes an input size and produces a generator
/// for new base maps. `f_patch` is a function which produces a map to be
/// applied to the base map.
fn map_apply<M, MBase, MPatch, F1, F2, F3>(
    group: &mut BenchmarkGroup<M>,
    bench_name: &str,
    f_base: F1,
    f_patch: F3,
) where
    M: Measurement,
    MBase: Map,
    MPatch: Map,
    F1: Fn(usize) -> F2,
    F2: Fn() -> MBase,
    F3: Fn(usize) -> MPatch,
{
    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));
        group.bench_with_input(
            BenchmarkId::new(bench_name, size),
            &size,
            |bencher, &size| {
                let f_base = f_base(size);
                bencher.iter_batched(
                    || (f_base(), f_patch(size)),
                    |(mut base, patch)| base.apply(black_box(&patch)),
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

fn concrete_map_apply(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("concrete_map_apply");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    let empty_base = |_: usize| || HashMap::<u64, u64>::default();

    let key_range_base = |size: usize| {
        move || {
            (0..size as u64)
                .zip(iter::repeat(0))
                .collect::<HashMap<u64, u64>>()
        }
    };

    let key_range_patch = |size: usize| {
        (0..size as u64)
            .zip(iter::repeat(1))
            .collect::<HashMap<u64, u64>>()
    };

    let disjoint_patch = |size: usize| {
        (size as u64..2 * size as u64)
            .zip(iter::repeat(1))
            .collect::<HashMap<u64, u64>>()
    };

    map_apply(
        &mut group,
        "empty_base_concrete_patch",
        empty_base,
        key_range_patch,
    );

    map_apply(&mut group, "empty_base_dynamic_patch", empty_base, |size| {
        key_range_patch(size).clone_dynamic()
    });

    map_apply(
        &mut group,
        "same_keys_concrete_patch",
        key_range_base,
        key_range_patch,
    );

    map_apply(
        &mut group,
        "same_keys_dynamic_patch",
        key_range_base,
        |size| key_range_patch(size).clone_dynamic(),
    );

    map_apply(
        &mut group,
        "disjoint_keys_concrete_patch",
        key_range_base,
        disjoint_patch,
    );

    map_apply(
        &mut group,
        "disjoint_keys_dynamic_patch",
        key_range_base,
        |size| disjoint_patch(size).clone_dynamic(),
    );
}

fn u64_to_n_byte_key(k: u64, n: usize) -> String {
    let mut key = String::with_capacity(n);
    write!(&mut key, "{}", k).unwrap();

    // Pad key to n bytes.
    key.extend(iter::repeat('\0').take(n - key.len()));
    key
}

fn dynamic_map_apply(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("dynamic_map_apply");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    let empty_base = |_: usize| || DynamicMap::default();

    let key_range_base = |size: usize| {
        move || {
            (0..size as u64)
                .zip(iter::repeat(0))
                .collect::<HashMap<u64, u64>>()
                .clone_dynamic()
        }
    };

    let key_range_patch = |size: usize| {
        (0..size as u64)
            .zip(iter::repeat(1))
            .collect::<HashMap<u64, u64>>()
    };

    let disjoint_patch = |size: usize| {
        (size as u64..2 * size as u64)
            .zip(iter::repeat(1))
            .collect::<HashMap<u64, u64>>()
    };

    map_apply(
        &mut group,
        "empty_base_concrete_patch",
        empty_base,
        key_range_patch,
    );

    map_apply(&mut group, "empty_base_dynamic_patch", empty_base, |size| {
        key_range_patch(size).clone_dynamic()
    });

    map_apply(
        &mut group,
        "same_keys_concrete_patch",
        key_range_base,
        key_range_patch,
    );

    map_apply(
        &mut group,
        "same_keys_dynamic_patch",
        key_range_base,
        |size| key_range_patch(size).clone_dynamic(),
    );

    map_apply(
        &mut group,
        "disjoint_keys_concrete_patch",
        key_range_base,
        disjoint_patch,
    );

    map_apply(
        &mut group,
        "disjoint_keys_dynamic_patch",
        key_range_base,
        |size| disjoint_patch(size).clone_dynamic(),
    );
}

fn dynamic_map_get(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("dynamic_map_get");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));
        group.bench_with_input(
            BenchmarkId::new("u64_keys", size),
            &size,
            |bencher, &size| {
                let mut map = DynamicMap::default();
                for i in 0..size as u64 {
                    map.insert(i, i);
                }

                bencher.iter(|| {
                    for i in 0..size as u64 {
                        let key = black_box(i);
                        black_box(assert!(map.get(&key).is_some()));
                    }
                });
            },
        );
    }

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));
        group.bench_with_input(
            BenchmarkId::new("64_byte_keys", size),
            &size,
            |bencher, &size| {
                let mut map = DynamicMap::default();
                let mut keys = Vec::with_capacity(size);
                for i in 0..size as u64 {
                    let key = u64_to_n_byte_key(i, 64);
                    map.insert(key.clone(), i);
                    keys.push(key);
                }

                bencher.iter(|| {
                    for i in 0..size {
                        let key = black_box(&keys[i]);
                        assert!(map.get(key).is_some());
                    }
                });
            },
        );
    }
}

fn dynamic_map_insert(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("dynamic_map_insert");
    group.warm_up_time(WARM_UP_TIME);
    group.measurement_time(MEASUREMENT_TIME);

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));
        group.bench_with_input(
            BenchmarkId::new("u64_keys", size),
            &size,
            |bencher, &size| {
                bencher.iter_batched(
                    || DynamicMap::default(),
                    |mut map| {
                        for i in 0..size as u64 {
                            let key = black_box(i);
                            black_box(map.insert(key, i));
                        }
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }

    for size in SIZES {
        group.throughput(Throughput::Elements(size as u64));
        group.bench_with_input(
            BenchmarkId::new("64_byte_keys", size),
            &size,
            |bencher, &size| {
                let mut keys = Vec::with_capacity(size);
                for i in 0..size {
                    let key = u64_to_n_byte_key(i as u64, 64);
                    keys.push(key);
                }

                bencher.iter_batched(
                    || (DynamicMap::default(), keys.clone()),
                    |(mut map, keys)| {
                        for (i, key) in keys.into_iter().enumerate() {
                            let key = black_box(key);
                            map.insert(key, i);
                        }
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }
}
