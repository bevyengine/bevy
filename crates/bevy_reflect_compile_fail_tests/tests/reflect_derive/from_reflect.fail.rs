use bevy_reflect::{FromReflect, Reflect};

// Reason: Cannot have conflicting `from_reflect` attributes
#[derive(Reflect)]
#[reflect(from_reflect = false)]
#[reflect(from_reflect = true)]
struct Foo {
    value: String,
}

// Reason: Cannot have conflicting `from_reflect` attributes
#[derive(Reflect)]
#[reflect(from_reflect = true)]
#[reflect(from_reflect = false)]
struct Bar {
    value: String,
}

// Reason: Conflicting `FromReflect` implementations
#[derive(Reflect, FromReflect)]
struct Baz {
    value: String,
}

fn main() {}
