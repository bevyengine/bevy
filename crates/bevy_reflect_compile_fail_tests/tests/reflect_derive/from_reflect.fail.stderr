error: `from_reflect` already set to false
 --> tests/reflect_derive/from_reflect.fail.rs:6:26
  |
6 | #[reflect(from_reflect = true)]
  |                          ^^^^

error: `from_reflect` already set to true
  --> tests/reflect_derive/from_reflect.fail.rs:14:26
   |
14 | #[reflect(from_reflect = false)]
   |                          ^^^^^

error[E0119]: conflicting implementations of trait `FromReflect` for type `Baz`
  --> tests/reflect_derive/from_reflect.fail.rs:20:19
   |
20 | #[derive(Reflect, FromReflect)]
   |          -------  ^^^^^^^^^^^ conflicting implementation for `Baz`
   |          |
   |          first implementation here
   |
   = note: this error originates in the derive macro `FromReflect` (in Nightly builds, run with -Z macro-backtrace for more info)
